<a href="https://colab.research.google.com/github/LucasSerranog/IA_udea/blob/main/02_preprocesado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Preprocesado de los datos**
#integrante
Lucas Serrano CC.1000944832


A continuacion se carga nuevamente el dataset

In [ ]:
!pip install -q kaggle

In [ ]:
from google.colab import files
files.upload()

Saving kaggle.json to kaggle.json


{'kaggle.json': b'{"username":"lucasserranogomez","key":"298bacde71f70e2178429e6df8c534fb"}'}

In [ ]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

In [ ]:
!kaggle competitions download -c udea-ai-4-eng-20251-pruebas-saber-pro-colombia

  0% 0.00/29.9M [00:00<?, ?B/s]
100% 29.9M/29.9M [00:00<00:00, 615MB/s]


In [ ]:
!unzip '/content/udea-ai-4-eng-20251-pruebas-saber-pro-colombia.zip'

Archive:  /content/udea-ai-4-eng-20251-pruebas-saber-pro-colombia.zip
  inflating: submission_example.csv  
  inflating: test.csv                
  inflating: train.csv               


In [ ]:
import pandas as pd  # Librería para manejo de dataframes
import seaborn as sns  # Librería para visualizaciones estadísticas
import numpy as np  # Librería para operaciones numéricas
import matplotlib.pyplot as plt  # Visualización con gráficos
from progressbar import progressbar as pbar  # Barra de progreso
from sklearn.pipeline import Pipeline
import sys



# Activa visualización en notebooks
%matplotlib inline


d = pd.read_csv(r'/content/train.csv')
d.head()


ID  PERIODO         ESTU_PRGM_ACADEMICO ESTU_PRGM_DEPARTAMENTO  \
0  904256    20212                  ENFERMERIA                 BOGOTÁ   
1  645256    20212                     DERECHO              ATLANTICO   
2  308367    20203       MERCADEO Y PUBLICIDAD                 BOGOTÁ   
3  470353    20195  ADMINISTRACION DE EMPRESAS              SANTANDER   
4  989032    20212                  PSICOLOGIA              ANTIOQUIA   

             ESTU_VALORMATRICULAUNIVERSIDAD ESTU_HORASSEMANATRABAJA  \
0  Entre 5.5 millones y menos de 7 millones       Menos de 10 horas   
1  Entre 2.5 millones y menos de 4 millones                       0   
2  Entre 2.5 millones y menos de 4 millones         Más de 30 horas   
3  Entre 4 millones y menos de 5.5 millones                       0   
4  Entre 2.5 millones y menos de 4 millones     Entre 21 y 30 horas   

  FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET                 FAMI_EDUCACIONPADRE  \
0            Estrato 3                 Si    Técnica o tecnológica incompleta   
1            Estrato 3                 No      Técnica o tecnológica completa   
2            Estrato 3                 Si  Secundaria (Bachillerato) completa   
3            Estrato 4                 Si                             No sabe   
4            Estrato 3                 Si                   Primaria completa   

  FAMI_TIENELAVADORA  ... ESTU_PRIVADO_LIBERTAD ESTU_PAGOMATRICULAPROPIO  \
0                 Si  ...                     N                       No   
1                 Si  ...                     N                       No   
2                 Si  ...                     N                       No   
3                 Si  ...                     N                       No   
4                 Si  ...                     N                       No   

  FAMI_TIENECOMPUTADOR FAMI_TIENEINTERNET.1  \
0                   Si                   Si   
1                   Si                   No   
2                   No                   Si   
3                   Si                   Si   
4                   Si                   Si   

                  FAMI_EDUCACIONMADRE RENDIMIENTO_GLOBAL coef_1  coef_2  \
0                           Postgrado         medio-alto  0.322   0.208   
1    Técnica o tecnológica incompleta               bajo  0.311   0.215   
2  Secundaria (Bachillerato) completa               bajo  0.297   0.214   
3  Secundaria (Bachillerato) completa               alto  0.485   0.172   
4                   Primaria completa         medio-bajo  0.316   0.232   

   coef_3  coef_4  
0   0.310   0.267  
1   0.292   0.264  
2   0.305   0.264  
3   0.252   0.190  
4   0.285   0.294  

[5 rows x 21 columns]

#limpieza y preprocesado de datos

Hacemos la columna **"id"** como indice de nuestro conjunto de datos

In [ ]:
if 'index' in d.columns:
    d.drop(columns=['index'], inplace=True)

# Establecer 'ID' como índice del DataFrame
d.set_index('ID', inplace=True)

# Verificar el resultado
d.head()

PERIODO         ESTU_PRGM_ACADEMICO ESTU_PRGM_DEPARTAMENTO  \
ID                                                                   
904256    20212                  ENFERMERIA                 BOGOTÁ   
645256    20212                     DERECHO              ATLANTICO   
308367    20203       MERCADEO Y PUBLICIDAD                 BOGOTÁ   
470353    20195  ADMINISTRACION DE EMPRESAS              SANTANDER   
989032    20212                  PSICOLOGIA              ANTIOQUIA   

                  ESTU_VALORMATRICULAUNIVERSIDAD ESTU_HORASSEMANATRABAJA  \
ID                                                                         
904256  Entre 5.5 millones y menos de 7 millones       Menos de 10 horas   
645256  Entre 2.5 millones y menos de 4 millones                       0   
308367  Entre 2.5 millones y menos de 4 millones         Más de 30 horas   
470353  Entre 4 millones y menos de 5.5 millones                       0   
989032  Entre 2.5 millones y menos de 4 millones     Entre 21 y 30 horas   

       FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET  \
ID                                               
904256            Estrato 3                 Si   
645256            Estrato 3                 No   
308367            Estrato 3                 Si   
470353            Estrato 4                 Si   
989032            Estrato 3                 Si   

                       FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  \
ID                                                              
904256    Técnica o tecnológica incompleta                 Si   
645256      Técnica o tecnológica completa                 Si   
308367  Secundaria (Bachillerato) completa                 Si   
470353                             No sabe                 Si   
989032                   Primaria completa                 Si   

       FAMI_TIENEAUTOMOVIL ESTU_PRIVADO_LIBERTAD ESTU_PAGOMATRICULAPROPIO  \
ID                                                                          
904256                  Si                     N                       No   
645256                  No                     N                       No   
308367                  No                     N                       No   
470353                  No                     N                       No   
989032                  Si                     N                       No   

       FAMI_TIENECOMPUTADOR FAMI_TIENEINTERNET.1  \
ID                                                 
904256                   Si                   Si   
645256                   Si                   No   
308367                   No                   Si   
470353                   Si                   Si   
989032                   Si                   Si   

                       FAMI_EDUCACIONMADRE RENDIMIENTO_GLOBAL  coef_1  coef_2  \
ID                                                                              
904256                           Postgrado         medio-alto   0.322   0.208   
645256    Técnica o tecnológica incompleta               bajo   0.311   0.215   
308367  Secundaria (Bachillerato) completa               bajo   0.297   0.214   
470353  Secundaria (Bachillerato) completa               alto   0.485   0.172   
989032                   Primaria completa         medio-bajo   0.316   0.232   

        coef_3  coef_4  
ID                      
904256   0.310   0.267  
645256   0.292   0.264  
308367   0.305   0.264  
470353   0.252   0.190  
989032   0.285   0.294

Eliminamos la columna **[FAMI_TIENEINTERNET.1	]** ya que son datos que se repiten en la columna **[FAMI_TIENEINTERNET]**

In [ ]:
d = d.drop(columns=['FAMI_TIENEINTERNET.1'], errors='ignore')
d.head()

PERIODO         ESTU_PRGM_ACADEMICO ESTU_PRGM_DEPARTAMENTO  \
ID                                                                   
904256    20212                  ENFERMERIA                 BOGOTÁ   
645256    20212                     DERECHO              ATLANTICO   
308367    20203       MERCADEO Y PUBLICIDAD                 BOGOTÁ   
470353    20195  ADMINISTRACION DE EMPRESAS              SANTANDER   
989032    20212                  PSICOLOGIA              ANTIOQUIA   

                  ESTU_VALORMATRICULAUNIVERSIDAD ESTU_HORASSEMANATRABAJA  \
ID                                                                         
904256  Entre 5.5 millones y menos de 7 millones       Menos de 10 horas   
645256  Entre 2.5 millones y menos de 4 millones                       0   
308367  Entre 2.5 millones y menos de 4 millones         Más de 30 horas   
470353  Entre 4 millones y menos de 5.5 millones                       0   
989032  Entre 2.5 millones y menos de 4 millones     Entre 21 y 30 horas   

       FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET  \
ID                                               
904256            Estrato 3                 Si   
645256            Estrato 3                 No   
308367            Estrato 3                 Si   
470353            Estrato 4                 Si   
989032            Estrato 3                 Si   

                       FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  \
ID                                                              
904256    Técnica o tecnológica incompleta                 Si   
645256      Técnica o tecnológica completa                 Si   
308367  Secundaria (Bachillerato) completa                 Si   
470353                             No sabe                 Si   
989032                   Primaria completa                 Si   

       FAMI_TIENEAUTOMOVIL ESTU_PRIVADO_LIBERTAD ESTU_PAGOMATRICULAPROPIO  \
ID                                                                          
904256                  Si                     N                       No   
645256                  No                     N                       No   
308367                  No                     N                       No   
470353                  No                     N                       No   
989032                  Si                     N                       No   

       FAMI_TIENECOMPUTADOR                 FAMI_EDUCACIONMADRE  \
ID                                                                
904256                   Si                           Postgrado   
645256                   Si    Técnica o tecnológica incompleta   
308367                   No  Secundaria (Bachillerato) completa   
470353                   Si  Secundaria (Bachillerato) completa   
989032                   Si                   Primaria completa   

       RENDIMIENTO_GLOBAL  coef_1  coef_2  coef_3  coef_4  
ID                                                         
904256         medio-alto   0.322   0.208   0.310   0.267  
645256               bajo   0.311   0.215   0.292   0.264  
308367               bajo   0.297   0.214   0.305   0.264  
470353               alto   0.485   0.172   0.252   0.190  
989032         medio-bajo   0.316   0.232   0.285   0.294

Ademas eliminamos la variable **[ESTU_PRIVADO_LIBERTAD]** ya que la cantidad de personas privadas de libertad son muy minimas y poco relevantes para nuestros datos a nivel general

In [ ]:
d.ESTU_PRIVADO_LIBERTAD = d.ESTU_PRIVADO_LIBERTAD.fillna('no info')
d.ESTU_PRIVADO_LIBERTAD.value_counts()

ESTU_PRIVADO_LIBERTAD
N    692466
S        34
Name: count, dtype: int64

Ademas de eliminar la columna **[Periodo]** ya que no es relevante o significativo para el analisis del rendimiento global de los estudantes

In [ ]:
d = d.drop(columns=['ESTU_PRIVADO_LIBERTAD'], errors='ignore')
d = d.drop(columns=['PERIODO'], errors='ignore')
d.head()

ESTU_PRGM_ACADEMICO ESTU_PRGM_DEPARTAMENTO  \
ID                                                          
904256                  ENFERMERIA                 BOGOTÁ   
645256                     DERECHO              ATLANTICO   
308367       MERCADEO Y PUBLICIDAD                 BOGOTÁ   
470353  ADMINISTRACION DE EMPRESAS              SANTANDER   
989032                  PSICOLOGIA              ANTIOQUIA   

                  ESTU_VALORMATRICULAUNIVERSIDAD ESTU_HORASSEMANATRABAJA  \
ID                                                                         
904256  Entre 5.5 millones y menos de 7 millones       Menos de 10 horas   
645256  Entre 2.5 millones y menos de 4 millones                       0   
308367  Entre 2.5 millones y menos de 4 millones         Más de 30 horas   
470353  Entre 4 millones y menos de 5.5 millones                       0   
989032  Entre 2.5 millones y menos de 4 millones     Entre 21 y 30 horas   

       FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET  \
ID                                               
904256            Estrato 3                 Si   
645256            Estrato 3                 No   
308367            Estrato 3                 Si   
470353            Estrato 4                 Si   
989032            Estrato 3                 Si   

                       FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  \
ID                                                              
904256    Técnica o tecnológica incompleta                 Si   
645256      Técnica o tecnológica completa                 Si   
308367  Secundaria (Bachillerato) completa                 Si   
470353                             No sabe                 Si   
989032                   Primaria completa                 Si   

       FAMI_TIENEAUTOMOVIL ESTU_PAGOMATRICULAPROPIO FAMI_TIENECOMPUTADOR  \
ID                                                                         
904256                  Si                       No                   Si   
645256                  No                       No                   Si   
308367                  No                       No                   No   
470353                  No                       No                   Si   
989032                  Si                       No                   Si   

                       FAMI_EDUCACIONMADRE RENDIMIENTO_GLOBAL  coef_1  coef_2  \
ID                                                                              
904256                           Postgrado         medio-alto   0.322   0.208   
645256    Técnica o tecnológica incompleta               bajo   0.311   0.215   
308367  Secundaria (Bachillerato) completa               bajo   0.297   0.214   
470353  Secundaria (Bachillerato) completa               alto   0.485   0.172   
989032                   Primaria completa         medio-bajo   0.316   0.232   

        coef_3  coef_4  
ID                      
904256   0.310   0.267  
645256   0.292   0.264  
308367   0.305   0.264  
470353   0.252   0.190  
989032   0.285   0.294

Hacemos el Conteo de estudiantes totales de la encuesta

In [ ]:
def contar_estudiantes(df):
    """
    Retorna la cantidad de estudiantes únicos en el DataFrame.
    Asume que cada estudiante tiene un ID único como índice.
    """
    return df.shape[0]
cantidad = contar_estudiantes(d)
print(f"Cantidad de estudiantes: {cantidad}")

Cantidad de estudiantes: 692500


Hacemos el análisis inicial de los valores faltantes de nuestro conjunto de datos **(NaN)** por cada una de las columnas de nuestro base de datos


In [ ]:
k=d.isna().sum()
k[k!=0]

ESTU_VALORMATRICULAUNIVERSIDAD     6287
ESTU_HORASSEMANATRABAJA           30857
FAMI_ESTRATOVIVIENDA              32137
FAMI_TIENEINTERNET                26629
FAMI_EDUCACIONPADRE               23178
FAMI_TIENELAVADORA                39773
FAMI_TIENEAUTOMOVIL               43623
ESTU_PAGOMATRICULAPROPIO           6498
FAMI_TIENECOMPUTADOR              38103
FAMI_EDUCACIONMADRE               23664
dtype: int64

ANALISIS: como el valor mayor de datos faltantes es de 43623 y los estudiantes a analizar son casi 700000 es decir un 6% de datos faltantes se imputan esos datos faltantes ya que son muy minimos o poco relevantes a comparacion de nuestra cantidad de datos totales

In [ ]:
#  Separar variables por tipo
cat_cols = d.select_dtypes(include='object').columns.tolist()
num_cols = d.select_dtypes(include=['int64', 'float64']).columns.tolist()

cat_cols, num_cols


(['ESTU_PRGM_ACADEMICO',
  'ESTU_PRGM_DEPARTAMENTO',
  'ESTU_VALORMATRICULAUNIVERSIDAD',
  'ESTU_HORASSEMANATRABAJA',
  'FAMI_ESTRATOVIVIENDA',
  'FAMI_TIENEINTERNET',
  'FAMI_EDUCACIONPADRE',
  'FAMI_TIENELAVADORA',
  'FAMI_TIENEAUTOMOVIL',
  'ESTU_PAGOMATRICULAPROPIO',
  'FAMI_TIENECOMPUTADOR',
  'FAMI_EDUCACIONMADRE',
  'RENDIMIENTO_GLOBAL'],
 ['coef_1', 'coef_2', 'coef_3', 'coef_4'])

In [ ]:
d.head()

ESTU_PRGM_ACADEMICO ESTU_PRGM_DEPARTAMENTO  \
ID                                                          
904256                  ENFERMERIA                 BOGOTÁ   
645256                     DERECHO              ATLANTICO   
308367       MERCADEO Y PUBLICIDAD                 BOGOTÁ   
470353  ADMINISTRACION DE EMPRESAS              SANTANDER   
989032                  PSICOLOGIA              ANTIOQUIA   

                  ESTU_VALORMATRICULAUNIVERSIDAD ESTU_HORASSEMANATRABAJA  \
ID                                                                         
904256  Entre 5.5 millones y menos de 7 millones       Menos de 10 horas   
645256  Entre 2.5 millones y menos de 4 millones                       0   
308367  Entre 2.5 millones y menos de 4 millones         Más de 30 horas   
470353  Entre 4 millones y menos de 5.5 millones                       0   
989032  Entre 2.5 millones y menos de 4 millones     Entre 21 y 30 horas   

       FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET  \
ID                                               
904256            Estrato 3                 Si   
645256            Estrato 3                 No   
308367            Estrato 3                 Si   
470353            Estrato 4                 Si   
989032            Estrato 3                 Si   

                       FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  \
ID                                                              
904256    Técnica o tecnológica incompleta                 Si   
645256      Técnica o tecnológica completa                 Si   
308367  Secundaria (Bachillerato) completa                 Si   
470353                             No sabe                 Si   
989032                   Primaria completa                 Si   

       FAMI_TIENEAUTOMOVIL ESTU_PAGOMATRICULAPROPIO FAMI_TIENECOMPUTADOR  \
ID                                                                         
904256                  Si                       No                   Si   
645256                  No                       No                   Si   
308367                  No                       No                   No   
470353                  No                       No                   Si   
989032                  Si                       No                   Si   

                       FAMI_EDUCACIONMADRE RENDIMIENTO_GLOBAL  coef_1  coef_2  \
ID                                                                              
904256                           Postgrado         medio-alto   0.322   0.208   
645256    Técnica o tecnológica incompleta               bajo   0.311   0.215   
308367  Secundaria (Bachillerato) completa               bajo   0.297   0.214   
470353  Secundaria (Bachillerato) completa               alto   0.485   0.172   
989032                   Primaria completa         medio-bajo   0.316   0.232   

        coef_3  coef_4  
ID                      
904256   0.310   0.267  
645256   0.292   0.264  
308367   0.305   0.264  
470353   0.252   0.190  
989032   0.285   0.294

limpiamos datos faltantes y luego analizar la frecuencia de respuestas

In [ ]:
d.FAMI_TIENEINTERNET = d.FAMI_TIENEINTERNET.fillna('no info')
d.FAMI_TIENEINTERNET.value_counts()

FAMI_TIENEINTERNET
Si         592514
No          73357
no info     26629
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.FAMI_TIENEINTERNET.values

# Obtener solo los valores únicos válidos (no NaN)
FAMI_TIENEINTERNET_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
FAMI_TIENEINTERNET_onehot_vals = {
    val: np.eye(len(FAMI_TIENEINTERNET_vals))[i] for i, val in enumerate(FAMI_TIENEINTERNET_vals)
}

# Crear codificación, respetando los NaN
FAMI_TIENEINTERNET_onehot_enc = np.full((len(x), len(FAMI_TIENEINTERNET_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        FAMI_TIENEINTERNET_onehot_enc[i] = FAMI_TIENEINTERNET_onehot_vals[val]

# Crear DataFrame con columnas
FAMI_TIENEINTERNET_df = pd.DataFrame(
    FAMI_TIENEINTERNET_onehot_enc,
    columns=[f"FAMI_TIENEINTERNET__{v}" for v in FAMI_TIENEINTERNET_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([FAMI_TIENEINTERNET_df, d.drop(columns=["FAMI_TIENEINTERNET"])], axis=1)



In [ ]:
d

FAMI_TIENEINTERNET__No  FAMI_TIENEINTERNET__Si  \
ID                                                       
904256                     0.0                     1.0   
645256                     1.0                     0.0   
308367                     0.0                     1.0   
470353                     0.0                     1.0   
989032                     0.0                     1.0   
...                        ...                     ...   
25096                      0.0                     1.0   
754213                     0.0                     1.0   
504185                     0.0                     1.0   
986620                     1.0                     0.0   
933374                     0.0                     1.0   

        FAMI_TIENEINTERNET__no info                  ESTU_PRGM_ACADEMICO  \
ID                                                                         
904256                          0.0                           ENFERMERIA   
645256                          0.0                              DERECHO   
308367                          0.0                MERCADEO Y PUBLICIDAD   
470353                          0.0           ADMINISTRACION DE EMPRESAS   
989032                          0.0                           PSICOLOGIA   
...                             ...                                  ...   
25096                           0.0                             BIOLOGIA   
754213                          0.0                           PSICOLOGIA   
504185                          0.0  ADMINISTRACIÓN EN SALUD OCUPACIONAL   
986620                          0.0                           PSICOLOGIA   
933374                          0.0                           PSICOLOGIA   

       ESTU_PRGM_DEPARTAMENTO            ESTU_VALORMATRICULAUNIVERSIDAD  \
ID                                                                        
904256                 BOGOTÁ  Entre 5.5 millones y menos de 7 millones   
645256              ATLANTICO  Entre 2.5 millones y menos de 4 millones   
308367                 BOGOTÁ  Entre 2.5 millones y menos de 4 millones   
470353              SANTANDER  Entre 4 millones y menos de 5.5 millones   
989032              ANTIOQUIA  Entre 2.5 millones y menos de 4 millones   
...                       ...                                       ...   
25096              LA GUAJIRA         Entre 500 mil y menos de 1 millón   
754213        NORTE SANTANDER  Entre 2.5 millones y menos de 4 millones   
504185                 BOGOTÁ    Entre 1 millón y menos de 2.5 millones   
986620                 TOLIMA  Entre 2.5 millones y menos de 4 millones   
933374              ANTIOQUIA  Entre 5.5 millones y menos de 7 millones   

       ESTU_HORASSEMANATRABAJA FAMI_ESTRATOVIVIENDA  \
ID                                                    
904256       Menos de 10 horas            Estrato 3   
645256                       0            Estrato 3   
308367         Más de 30 horas            Estrato 3   
470353                       0            Estrato 4   
989032     Entre 21 y 30 horas            Estrato 3   
...                        ...                  ...   
25096      Entre 11 y 20 horas            Estrato 2   
754213         Más de 30 horas            Estrato 3   
504185       Menos de 10 horas            Estrato 3   
986620       Menos de 10 horas            Estrato 1   
933374         Más de 30 horas            Estrato 4   

                       FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  \
ID                                                              
904256    Técnica o tecnológica incompleta                 Si   
645256      Técnica o tecnológica completa                 Si   
308367  Secundaria (Bachillerato) completa                 Si   
470353                             No sabe                 Si   
989032                   Primaria completa                 Si   
...                                    ...                ...   
25096   Secundaria (Bachillerato) completa          

repetimos para el resto de datos

In [ ]:
d.ESTU_VALORMATRICULAUNIVERSIDAD = d.ESTU_VALORMATRICULAUNIVERSIDAD.fillna('no info')
d.ESTU_VALORMATRICULAUNIVERSIDAD.value_counts()

ESTU_VALORMATRICULAUNIVERSIDAD
Entre 1 millón y menos de 2.5 millones      204048
Entre 2.5 millones y menos de 4 millones    127430
Menos de 500 mil                             80263
Entre 500 mil y menos de 1 millón            78704
Entre 4 millones y menos de 5.5 millones     69736
Más de 7 millones                            68014
Entre 5.5 millones y menos de 7 millones     38490
No pagó matrícula                            19528
no info                                       6287
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.ESTU_VALORMATRICULAUNIVERSIDAD.values

# Obtener solo los valores únicos válidos (no NaN)
ESTU_VALORMATRICULAUNIVERSIDAD_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
ESTU_VALORMATRICULAUNIVERSIDAD_onehot_vals = {
    val: np.eye(len(ESTU_VALORMATRICULAUNIVERSIDAD_vals))[i] for i, val in enumerate(ESTU_VALORMATRICULAUNIVERSIDAD_vals)
}

# Crear codificación, respetando los NaN
ESTU_VALORMATRICULAUNIVERSIDAD_onehot_enc = np.full((len(x), len(ESTU_VALORMATRICULAUNIVERSIDAD_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        ESTU_VALORMATRICULAUNIVERSIDAD_onehot_enc[i] = ESTU_VALORMATRICULAUNIVERSIDAD_onehot_vals[val]

# Crear DataFrame con columnas
ESTU_VALORMATRICULAUNIVERSIDAD_df = pd.DataFrame(
    ESTU_VALORMATRICULAUNIVERSIDAD_onehot_enc,
    columns=[f"ESTU_VALORMATRICULAUNIVERSIDAD__{v}" for v in ESTU_VALORMATRICULAUNIVERSIDAD_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([ESTU_VALORMATRICULAUNIVERSIDAD_df, d.drop(columns=["ESTU_VALORMATRICULAUNIVERSIDAD"])], axis=1)

In [ ]:
d.ESTU_HORASSEMANATRABAJA = d.ESTU_HORASSEMANATRABAJA.fillna('no info')
d.ESTU_HORASSEMANATRABAJA.value_counts()

ESTU_HORASSEMANATRABAJA
Más de 30 horas        249352
0                      116550
Entre 11 y 20 horas    115857
Entre 21 y 30 horas     92693
Menos de 10 horas       87191
no info                 30857
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.ESTU_HORASSEMANATRABAJA.values

# Obtener solo los valores únicos válidos (no NaN)
ESTU_HORASSEMANATRABAJA_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
ESTU_HORASSEMANATRABAJA_onehot_vals = {
    val: np.eye(len(ESTU_HORASSEMANATRABAJA_vals))[i] for i, val in enumerate(ESTU_HORASSEMANATRABAJA_vals)
}

# Crear codificación, respetando los NaN
ESTU_HORASSEMANATRABAJA_onehot_enc = np.full((len(x), len(ESTU_HORASSEMANATRABAJA_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        ESTU_HORASSEMANATRABAJA_onehot_enc[i] = ESTU_HORASSEMANATRABAJA_onehot_vals[val]

# Crear DataFrame con columnas
ESTU_HORASSEMANATRABAJA_df = pd.DataFrame(
    ESTU_HORASSEMANATRABAJA_onehot_enc,
    columns=[f"ESTU_HORASSEMANATRABAJA__{v}" for v in ESTU_HORASSEMANATRABAJA_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([ESTU_HORASSEMANATRABAJA_df, d.drop(columns=["ESTU_HORASSEMANATRABAJA"])], axis=1)

In [ ]:
d.FAMI_ESTRATOVIVIENDA = d.FAMI_ESTRATOVIVIENDA.fillna('no info')
d.FAMI_ESTRATOVIVIENDA.value_counts()

FAMI_ESTRATOVIVIENDA
Estrato 2      232671
Estrato 3      210685
Estrato 1      111991
Estrato 4       65514
no info         32137
Estrato 5       23608
Estrato 6       12605
Sin Estrato      3289
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.FAMI_ESTRATOVIVIENDA.values

# Obtener solo los valores únicos válidos (no NaN)
FAMI_ESTRATOVIVIENDA_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
FAMI_ESTRATOVIVIENDA_onehot_vals = {
    val: np.eye(len(FAMI_ESTRATOVIVIENDA_vals))[i] for i, val in enumerate(FAMI_ESTRATOVIVIENDA_vals)
}

# Crear codificación, respetando los NaN
FAMI_ESTRATOVIVIENDA_onehot_enc = np.full((len(x), len(FAMI_ESTRATOVIVIENDA_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        FAMI_ESTRATOVIVIENDA_onehot_enc[i] = FAMI_ESTRATOVIVIENDA_onehot_vals[val]

# Crear DataFrame con columnas
FAMI_ESTRATOVIVIENDA_df = pd.DataFrame(
    FAMI_ESTRATOVIVIENDA_onehot_enc,
    columns=[f"FAMI_ESTRATOVIVIENDA__{v}" for v in FAMI_ESTRATOVIVIENDA_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([FAMI_ESTRATOVIVIENDA_df, d.drop(columns=["FAMI_ESTRATOVIVIENDA"])], axis=1)

In [ ]:
d.FAMI_EDUCACIONPADRE = [
    'no info' if pd.isna(i) or i in ['No sabe', 'No Aplica'] else i
    for i in d.FAMI_EDUCACIONPADRE.values
]

d.FAMI_EDUCACIONPADRE.value_counts()

FAMI_EDUCACIONPADRE
Secundaria (Bachillerato) completa      128289
Primaria incompleta                     125675
Educación profesional completa           83117
Secundaria (Bachillerato) incompleta     71654
Técnica o tecnológica completa           62995
Primaria completa                        55958
no info                                  48999
Postgrado                                44169
Educación profesional incompleta         27084
Técnica o tecnológica incompleta         22552
Ninguno                                  22008
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.FAMI_EDUCACIONPADRE.values

# Obtener solo los valores únicos válidos (no NaN)
FAMI_EDUCACIONPADRE_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
FAMI_EDUCACIONPADRE_onehot_vals = {
    val: np.eye(len(FAMI_EDUCACIONPADRE_vals))[i] for i, val in enumerate(FAMI_EDUCACIONPADRE_vals)
}

# Crear codificación, respetando los NaN
FAMI_EDUCACIONPADRE_onehot_enc = np.full((len(x), len(FAMI_EDUCACIONPADRE_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        FAMI_EDUCACIONPADRE_onehot_enc[i] = FAMI_EDUCACIONPADRE_onehot_vals[val]

# Crear DataFrame con columnas
FAMI_EDUCACIONPADRE_df = pd.DataFrame(
    FAMI_EDUCACIONPADRE_onehot_enc,
    columns=[f"FAMI_EDUCACIONPADRE__{v}" for v in FAMI_EDUCACIONPADRE_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([FAMI_EDUCACIONPADRE_df, d.drop(columns=["FAMI_EDUCACIONPADRE"])], axis=1)

In [ ]:
d.FAMI_TIENELAVADORA = d.FAMI_TIENELAVADORA.fillna('no info')
d.FAMI_TIENELAVADORA.value_counts()

FAMI_TIENELAVADORA
Si         563390
No          89337
no info     39773
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.FAMI_TIENELAVADORA.values

# Obtener solo los valores únicos válidos (no NaN)
FAMI_TIENELAVADORA_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
FAMI_TIENELAVADORA_onehot_vals = {
    val: np.eye(len(FAMI_TIENELAVADORA_vals))[i] for i, val in enumerate(FAMI_TIENELAVADORA_vals)
}

# Crear codificación, respetando los NaN
FAMI_TIENELAVADORA_onehot_enc = np.full((len(x), len(FAMI_TIENELAVADORA_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        FAMI_TIENELAVADORA_onehot_enc[i] = FAMI_TIENELAVADORA_onehot_vals[val]

# Crear DataFrame con columnas
FAMI_TIENELAVADORA_df = pd.DataFrame(
    FAMI_TIENELAVADORA_onehot_enc,
    columns=[f"FAMI_TIENELAVADORA__{v}" for v in FAMI_TIENELAVADORA_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([FAMI_TIENELAVADORA_df, d.drop(columns=["FAMI_TIENELAVADORA"])], axis=1)

In [ ]:
d.FAMI_TIENEAUTOMOVIL = d.FAMI_TIENEAUTOMOVIL.fillna('no info')
d.FAMI_TIENEAUTOMOVIL.value_counts()

FAMI_TIENEAUTOMOVIL
No         412606
Si         236271
no info     43623
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.FAMI_TIENEAUTOMOVIL.values

# Obtener solo los valores únicos válidos (no NaN)
FAMI_TIENEAUTOMOVIL_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
FAMI_TIENEAUTOMOVIL_onehot_vals = {
    val: np.eye(len(FAMI_TIENEAUTOMOVIL_vals))[i] for i, val in enumerate(FAMI_TIENEAUTOMOVIL_vals)
}

# Crear codificación, respetando los NaN
FAMI_TIENEAUTOMOVIL_onehot_enc = np.full((len(x), len(FAMI_TIENEAUTOMOVIL_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        FAMI_TIENEAUTOMOVIL_onehot_enc[i] = FAMI_TIENEAUTOMOVIL_onehot_vals[val]

# Crear DataFrame con columnas
FAMI_TIENEAUTOMOVIL_df = pd.DataFrame(
    FAMI_TIENEAUTOMOVIL_onehot_enc,
    columns=[f"FAMI_TIENEAUTOMOVIL__{v}" for v in FAMI_TIENEAUTOMOVIL_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([FAMI_TIENEAUTOMOVIL_df, d.drop(columns=["FAMI_TIENEAUTOMOVIL"])], axis=1)

In [ ]:
d.ESTU_PAGOMATRICULAPROPIO = d.ESTU_PAGOMATRICULAPROPIO.fillna('no info')
d.ESTU_PAGOMATRICULAPROPIO.value_counts()

ESTU_PAGOMATRICULAPROPIO
No         382201
Si         303801
no info      6498
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.ESTU_PAGOMATRICULAPROPIO.values

# Obtener solo los valores únicos válidos (no NaN)
ESTU_PAGOMATRICULAPROPIO_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
ESTU_PAGOMATRICULAPROPIO_onehot_vals = {
    val: np.eye(len(ESTU_PAGOMATRICULAPROPIO_vals))[i] for i, val in enumerate(ESTU_PAGOMATRICULAPROPIO_vals)
}

# Crear codificación, respetando los NaN
ESTU_PAGOMATRICULAPROPIO_onehot_enc = np.full((len(x), len(ESTU_PAGOMATRICULAPROPIO_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        ESTU_PAGOMATRICULAPROPIO_onehot_enc[i] = ESTU_PAGOMATRICULAPROPIO_onehot_vals[val]

# Crear DataFrame con columnas
ESTU_PAGOMATRICULAPROPIO_df = pd.DataFrame(
    ESTU_PAGOMATRICULAPROPIO_onehot_enc,
    columns=[f"ESTU_PAGOMATRICULAPROPIO__{v}" for v in ESTU_PAGOMATRICULAPROPIO_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([ESTU_PAGOMATRICULAPROPIO_df, d.drop(columns=["ESTU_PAGOMATRICULAPROPIO"])], axis=1)

In [ ]:
d.FAMI_TIENECOMPUTADOR = d.FAMI_TIENECOMPUTADOR.fillna('no info')

d.FAMI_TIENECOMPUTADOR.value_counts()

FAMI_TIENECOMPUTADOR
Si         597670
No          56727
no info     38103
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.FAMI_TIENECOMPUTADOR.values

# Obtener solo los valores únicos válidos (no NaN)
FAMI_TIENECOMPUTADOR_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
FAMI_TIENECOMPUTADOR_onehot_vals = {
    val: np.eye(len(FAMI_TIENECOMPUTADOR_vals))[i] for i, val in enumerate(FAMI_TIENECOMPUTADOR_vals)
}

# Crear codificación, respetando los NaN
FAMI_TIENECOMPUTADOR_onehot_enc = np.full((len(x), len(FAMI_TIENECOMPUTADOR_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        FAMI_TIENECOMPUTADOR_onehot_enc[i] = FAMI_TIENECOMPUTADOR_onehot_vals[val]

# Crear DataFrame con columnas
FAMI_TIENECOMPUTADOR_df = pd.DataFrame(
    FAMI_TIENECOMPUTADOR_onehot_enc,
    columns=[f"FAMI_TIENECOMPUTADOR__{v}" for v in FAMI_TIENECOMPUTADOR_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([FAMI_TIENECOMPUTADOR_df, d.drop(columns=["FAMI_TIENECOMPUTADOR"])], axis=1)

In [ ]:
d.FAMI_EDUCACIONMADRE = [
    'no info' if pd.isna(i) or i in ['No sabe', 'No Aplica'] else i
    for i in d.FAMI_EDUCACIONMADRE.values
]

d.FAMI_EDUCACIONMADRE.value_counts()

FAMI_EDUCACIONMADRE
Secundaria (Bachillerato) completa      141744
Primaria incompleta                      99420
Técnica o tecnológica completa           89542
Educación profesional completa           85326
Secundaria (Bachillerato) incompleta     81012
Primaria completa                        56125
Postgrado                                46246
no info                                  28599
Técnica o tecnológica incompleta         27533
Educación profesional incompleta         22470
Ninguno                                  14483
Name: count, dtype: int64

In [ ]:
# Extraer la columna original sin convertir a string
x = d.FAMI_EDUCACIONMADRE.values

# Obtener solo los valores únicos válidos (no NaN)
FAMI_EDUCACIONMADRE_vals = sorted(pd.Series(x).dropna().unique())

# Crear diccionario para codificación one-hot
FAMI_EDUCACIONMADRE_onehot_vals = {
    val: np.eye(len(FAMI_EDUCACIONMADRE_vals))[i] for i, val in enumerate(FAMI_EDUCACIONMADRE_vals)
}

# Crear codificación, respetando los NaN
FAMI_EDUCACIONMADRE_onehot_enc = np.full((len(x), len(FAMI_EDUCACIONMADRE_vals)), np.nan)
for i, val in enumerate(x):
    if pd.notna(val):
        FAMI_EDUCACIONMADRE_onehot_enc[i] = FAMI_EDUCACIONMADRE_onehot_vals[val]

# Crear DataFrame con columnas
FAMI_EDUCACIONMADRE_df = pd.DataFrame(
    FAMI_EDUCACIONMADRE_onehot_enc,
    columns=[f"FAMI_EDUCACIONMADRE__{v}" for v in FAMI_EDUCACIONMADRE_vals],
    index=d.index
)

# Concatenar al dataframe original
d = pd.concat([FAMI_EDUCACIONMADRE_df, d.drop(columns=["FAMI_EDUCACIONMADRE"])], axis=1)

eliminamos datos de las columnas **[no info]** ya que estos datos faltantes no son relevantes para nuestra analisis como dijimos anteriormente

In [ ]:
d = d[[col for col in d.columns if not col.endswith('no info')]]


Por ultimo hacemos un analisis por cada uno de nuestros datos y vemos como quedo nuestra base de datos con la limpieza que le realizamos

In [ ]:
d

FAMI_EDUCACIONMADRE__Educación profesional completa  \
ID                                                            
904256                                                0.0     
645256                                                0.0     
308367                                                0.0     
470353                                                0.0     
989032                                                0.0     
...                                                   ...     
25096                                                 0.0     
754213                                                0.0     
504185                                                0.0     
986620                                                0.0     
933374                                                0.0     

        FAMI_EDUCACIONMADRE__Educación profesional incompleta  \
ID                                                              
904256                                                0.0       
645256                                                0.0       
308367                                                0.0       
470353                                                0.0       
989032                                                0.0       
...                                                   ...       
25096                                                 0.0       
754213                                                0.0       
504185                                                0.0       
986620                                                0.0       
933374                                                0.0       

        FAMI_EDUCACIONMADRE__Ninguno  FAMI_EDUCACIONMADRE__Postgrado  \
ID                                                                     
904256                           0.0                             1.0   
645256                           0.0                             0.0   
308367                           0.0                             0.0   
470353                           0.0                             0.0   
989032                           0.0                             0.0   
...                              ...                             ...   
25096                            0.0                             0.0   
754213                           0.0                             0.0   
504185                           0.0                             0.0   
986620                           0.0                             0.0   
933374                           0.0                             0.0   

        FAMI_EDUCACIONMADRE__Primaria completa  \
ID                                               
904256                                     0.0   
645256                                     0.0   
308367                                     0.0   
470353                                     0.0   
989032                                     1.0   
...                                        ...   
25096                                      0.0   
754213                                     0.0   
504185                                     0.0   
986620                                     1.0   
933374                                     0.0   

        FAMI_EDUCACIONMADRE__Primaria incompleta  \
ID                                                 
904256                                       0.0   
645256                                       0.0   
308367                                       0.0   
470353                                       0.0   
989032                                       0.0   
...                                          ...   
25096                                        0.0   
754213                                       0.0   
504185                                       0.0   
986620                                       0.0   
933374                                       0.0   

        FAMI_EDUCACIONMADRE__Secundaria (Bachillerato) completa  \
ID        

In [ ]:
d.describe()

FAMI_EDUCACIONMADRE__Educación profesional completa  \
count                                      692500.000000     
mean                                            0.123214     
std                                             0.328683     
min                                             0.000000     
25%                                             0.000000     
50%                                             0.000000     
75%                                             0.000000     
max                                             1.000000     

       FAMI_EDUCACIONMADRE__Educación profesional incompleta  \
count                                      692500.000000       
mean                                            0.032448       
std                                             0.177186       
min                                             0.000000       
25%                                             0.000000       
50%                                             0.000000       
75%                                             0.000000       
max                                             1.000000       

       FAMI_EDUCACIONMADRE__Ninguno  FAMI_EDUCACIONMADRE__Postgrado  \
count                 692500.000000                   692500.000000   
mean                       0.020914                        0.066781   
std                        0.143097                        0.249643   
min                        0.000000                        0.000000   
25%                        0.000000                        0.000000   
50%                        0.000000                        0.000000   
75%                        0.000000                        0.000000   
max                        1.000000                        1.000000   

       FAMI_EDUCACIONMADRE__Primaria completa  \
count                           692500.000000   
mean                                 0.081047   
std                                  0.272907   
min                                  0.000000   
25%                                  0.000000   
50%                                  0.000000   
75%                                  0.000000   
max                                  1.000000   

       FAMI_EDUCACIONMADRE__Primaria incompleta  \
count                             692500.000000   
mean                                   0.143567   
std                                    0.350650   
min                                    0.000000   
25%                                    0.000000   
50%                                    0.000000   
75%                                    0.000000   
max                                    1.000000   

       FAMI_EDUCACIONMADRE__Secundaria (Bachillerato) completa  \
count                                      692500.000000         
mean                                            0.204684         
std                                             0.403471         
min                                             0.000000         
25%                                             0.000000         
50%                                             0.000000         
75%                                             0.000000         
max                                             1.000000         

       FAMI_EDUCACIONMADRE__Secundaria (Bachillerato) incompleta  \
count                                      692500.000000           
mean                                            0.116985           
std                                             0.321402           
min                                             0.000000           
25%                                             0.000000           
50%                                             0.000000           
75%                                             0.000000           
max                                             1.000000           

       FAMI_EDUCACIONMADRE__Técnica o tecnológica completa  \
count                                      692500.000000   

In [ ]:
# Guardar el dataset preprocesado para usarlo en los modelos
d.to_csv("datos_preprocesados.csv", index=False)

print("✅ Dataset preprocesado guardado como 'datos_preprocesados.csv'")


✅ Dataset preprocesado guardado como 'datos_preprocesados.csv'


In [ ]:
import pandas as pd
pd.read_csv("datos_preprocesados.csv").head()


FAMI_EDUCACIONMADRE__Educación profesional completa  \
0                                                0.0     
1                                                0.0     
2                                                0.0     
3                                                0.0     
4                                                0.0     

   FAMI_EDUCACIONMADRE__Educación profesional incompleta  \
0                                                0.0       
1                                                0.0       
2                                                0.0       
3                                                0.0       
4                                                0.0       

   FAMI_EDUCACIONMADRE__Ninguno  FAMI_EDUCACIONMADRE__Postgrado  \
0                           0.0                             1.0   
1                           0.0                             0.0   
2                           0.0                             0.0   
3                           0.0                             0.0   
4                           0.0                             0.0   

   FAMI_EDUCACIONMADRE__Primaria completa  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     1.0   

   FAMI_EDUCACIONMADRE__Primaria incompleta  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   

   FAMI_EDUCACIONMADRE__Secundaria (Bachillerato) completa  \
0                                                0.0         
1                                                0.0         
2                                                1.0         
3                                                1.0         
4                                                0.0         

   FAMI_EDUCACIONMADRE__Secundaria (Bachillerato) incompleta  \
0                                                0.0           
1                                                0.0           
2                                                0.0           
3                                                0.0           
4                                                0.0           

   FAMI_EDUCACIONMADRE__Técnica o tecnológica completa  \
0                                                0.0     
1                                                0.0     
2                                                0.0     
3                                                0.0     
4                                                0.0     

   FAMI_EDUCACIONMADRE__Técnica o tecnológica incompleta  ...  \
0                                                0.0      ...   
1                                                1.0      ...   
2                                                0.0      ...   
3                                                0.0      ...   
4                                                0.0      ...   

   ESTU_VALORMATRICULAUNIVERSIDAD__No pagó matrícula  FAMI_TIENEINTERNET__No  \
0                                                0.0                     0.0   
1                                                0.0                     1.0   
2                                                0.0                     0.0   
3                                                0.0                     0.0   
4                                                0.0                     0.0   

   FAMI_TIENEINTERNET__Si         ESTU_PRGM_ACADEMICO  ESTU_PRGM_DEPARTAMENTO  \
0                     1.0                  ENFERMERIA                  BOGOTÁ   
1                     0.0                     DERECHO               ATLANTICO   
2                     1.0       MERCADEO Y PUBLICIDAD                  BOGOTÁ   
3                     1.0  ADMINI

In [ ]:
from google.colab import files
files.download("datos_preprocesados.csv")


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>